### PR TITLE
Using new method to deploy the programs

### PR DESCRIPTION
**Pull request by godmodegalactus**
#### Problem
 Deploying program uses old and inefficient method.

#### Summary of Changes
Using newer method to deploy programs.

The transactions are sent and confirmed in parallel.
The transactions are signed just before sending so they are more likely to get confirmed.

Depending on size of the program it takes much less time. Previously if the program was too big it was nearly impossible to deploy with CLI now deployments of even large programs should work efficiently.

Fixes #

### CODE REVIEW

**Review by KirillLykov — CHANGES_REQUESTED** [review]
Pull request description doesn't include the on how it is connected with previous PRs and what is relation with https://github.com/solana-labs/solana/pull/32631

As for [aforementioned PR](https://github.com/solana-labs/solana/pull/32631#issuecomment-1651743914), I would like to see some numbers for the program deployment.

**Comment by godmodegalactus** [comment]
Here is demo before and after this PR: https://www.youtube.com/watch?v=aNsZHmjsCFA

**Comment by godmodegalactus** [comment]
https://github.com/solana-labs/solana/issues/32595#issuecomment-1653941590

This PR will take little more time than 25 secs because we still send transactions over TPU, but is much faster and reliable than the previous implementation. 

https://github.com/solana-labs/solana/pull/32631#issuecomment-1651743914 This PR will make RPC as default mode to deploy programs.

File: client/src/send_and_confirm_transactions_in_parallel.rs [hunk 1]
```diff
@@ -250,6 +250,8 @@ async fn sign_all_messages_and_send<T: Signers + ?Sized>(
                 ),
             );
         }
+
+        tokio::time::sleep(Duration::from_millis(500)).await;
```
Comment (t-nelson) [inline client/src/send_and_confirm_transactions_in_parallel.rs:254]: looks unrelated to the primary change?

File: cli/src/program.rs [hunk 2]
```diff
@@ -2178,16 +2181,29 @@ fn send_deploy_messages(
...
+                        .block_on(tpu_client_fut)
+                        .expect("Should return a valid tpu client");
+
+                    send_and_confirm_transactions_in_parallel_blocking(
+                        rpc_client.clone(),
+                        Some(tpu_client),
+                        write_messages,
+                        &[payer_signer, write_signer],
+                        SendAndConfrimConfig {
+                            resign_txs_count: Some(5),
```
Comment (t-nelson) [inline cli/src/program.rs:2202]: maybe expose as a cli arg `--transaction-resign-attempts NUM_ATTEMPTS` default `0` for saftety?

File: client/src/send_and_confirm_transactions_in_parallel.rs (hunk 1 above)
Comment (godmodegalactus) [reply client/src/send_and_confirm_transactions_in_parallel.rs:254]: `program::tests::test_cli_keypair_file` are failing because of tokio scheduling. This is because in  tests tokio will be stuck in loop on this  task, and cannot schedule other tokio tasks.

File: cli/src/program.rs (hunk 2 above)
Comment (godmodegalactus) [reply cli/src/program.rs:2202]: I will update this in the other PR, for this PR it is better that we have the same number of resign attempts like the previous implementation.

File: client/src/send_and_confirm_transactions_in_parallel.rs (hunk 1 above)
Comment (t-nelson) [reply client/src/send_and_confirm_transactions_in_parallel.rs:254]: :thinking: 

sounds broken :detective:

Comment (godmodegalactus) [reply client/src/send_and_confirm_transactions_in_parallel.rs:254]: Not really, in real scenario tokio will wait for rpc_client to finish i/o request from jsonrpc client. So tokio is able to schedule tasks correctly. In test this rpc_client is mocked and so there is no real i/o that is why the task looped and finished all its tries. So this was not a real problem, it is just because the rpc_client in test is mocked and not really doing any i/o.

Comment (godmodegalactus) [reply client/src/send_and_confirm_transactions_in_parallel.rs:254]: This change will also reduce load on the RPC, it will wait at least 500ms before again sending the request `get_signature_status` to the RPC.

Comment (godmodegalactus) [reply client/src/send_and_confirm_transactions_in_parallel.rs]: Ok update mock rpc client to represent latency between rpc server and rpc client.

Comment (godmodegalactus) [reply client/src/send_and_confirm_transactions_in_parallel.rs]: Fixed and recommited.

Comment (godmodegalactus) [reply client/src/send_and_confirm_transactions_in_parallel.rs]: After some more analysis : https://github.com/solana-labs/solana/issues/32792 bug. Rpc client is no longer used incase of failure to send over TPU client. Currently, there is no way around a small wait to handle scheduling correctly.
I can just wait for 1ms to help tokio schedule the tasks correctly.

File: rpc-client/src/mock_sender.rs [hunk 3]
```diff
@@ -99,6 +99,8 @@ impl RpcSender for MockSender {
         request: RpcRequest,
         params: serde_json::Value,
     ) -> Result<serde_json::Value> {
+        // a small sleep to mock latency for io between rpc server and client
+        tokio::time::sleep(Duration::from_millis(1)).await;
```
Comment (t-nelson) [inline rpc-client/src/mock_sender.rs:103]: this is breaking the doctests which i guess don't have an async executor. it's the first introduction of `await` in the mock client, so presumably that. i think we'll need a wrapped sync version of `MockSender`, or yolo a `std::thread::sleep()` in and 🙈

**Review by t-nelson — COMMENTED** [review]
seeing this failure still
```
thread 'program::tests::test_cli_keypair_file' panicked at 'called `Result::unwrap()` on an `Err` value: "Data writes to account failed: Custom error: Max retries exceeded"', cli/src/program.rs:3165:56
```

**Comment by godmodegalactus** [comment]
> seeing this failure still
> 
> ```
> thread 'program::tests::test_cli_keypair_file' panicked at 'called `Result::unwrap()` on an `Err` value: "Data writes to account failed: Custom error: Max retries exceeded"', cli/src/program.rs:3165:56
> ```

Yes I see it. I have to spend more time to debug the issue.

**Comment by godmodegalactus** [comment]
@t-nelson approve?

File: client/src/send_and_confirm_transactions_in_parallel.rs [hunk 4]
```diff
@@ -286,7 +286,7 @@ async fn confirm_transactions_till_block_height_and_resend_unexpired_transaction
 
         // wait till all transactions are confirmed or we have surpassed max processing age for the last sent transaction
         while !unconfirmed_transasction_map.is_empty()
-            && current_block_height.load(Ordering::Relaxed) < max_valid_block_height
+            && current_block_height.load(Ordering::Relaxed) <= max_valid_block_height
```
Comment (t-nelson) [inline client/src/send_and_confirm_transactions_in_parallel.rs:289]: shouldn't really make unrelated changes like this. better to keep pr's tight and on topic to ease cases where backport or revert are applicable

**Review by t-nelson — APPROVED** [review]

**Comment by t-nelson** [comment]
@KirillLykov i think we need your :+1: 'cause of outstanding change request

**Review by KirillLykov — APPROVED** [review]

**Review by t-nelson — APPROVED** [review]
